pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -45,87 +45,49 @@ func init() {
-// same node, or multiple similar nodes if the pods exceed one node's capacity.
-func (s *Scheduler) getSchedules(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*Schedule {
-	schedules := []*Schedule{}
+// schedule separates pods into a list of TheoreticalNodes that contain the pods. All pods in each theoretical node
+// contain isomorphic scheduling constraints and can be deployed together on the same node, or multiple similar nodes if
+// the pods exceed one node's capacity.
+func (s *Scheduler) schedule(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*TheoreticalNode {
+	var nodes []*TheoreticalNode
 	for _, pod := range pods {
-		isCompatible := false
-		for _, schedule := range schedules {
-			if err := schedule.Requirements.Compatible(v1alpha5.NewPodRequirements(pod)); err == nil {
-				// Test if there is any instance type that can support the combined constraints
-				// TODO: Implement a virtual node approach solution that combine scheduling and node selection to solve this problem.
-				c := schedule.Tighten(pod)
-				for _, instanceType := range instanceTypes {
-					if support(instanceType, c) {
-						schedule.Constraints = c
-						schedule.Pods = append(schedule.Pods, pod)
-						isCompatible = true
-						break
-					}
-				}
+		canScheduleOnExistingNode := false
+		for _, node := range nodes {
+			if err := node.Compatible(pod); err == nil {
+				node.Add(pod)
+				canScheduleOnExistingNode = true
+				break
 			}
 		}
-		if !isCompatible {
-			schedules = append(schedules, &Schedule{Constraints: constraints.Tighten(pod), Pods: []*v1.Pod{pod}})
-		}
-	}
-	return schedules
-}
-
-func support(instanceType cloudprovider.InstanceType, constraints *v1alpha5.Constraints) bool {
-	req := []v1.NodeSelectorRequirement{}
-	for key := range constraints.Requirements.Keys() {
-		req = append(req, v1.NodeSelectorRequirement{Key: key, Operator: v1.NodeSelectorOpExists})
-	}
-	for _, offering := range instanceType.Offerings() {
-		supported := map[string][]string{}
-		supported[v1.LabelTopologyZone] = []string{offering.Zone}
-		supported[v1alpha5.LabelCapacityType] = []string{offering.CapacityType}
-		supported[v1.LabelInstanceTypeStable] = []string{instanceType.Name()}
-		supported[v1.LabelArchStable] = []string{instanceType.Architecture()}
-		supported[v1.LabelOSStable] = instanceType.OperatingSystems().UnsortedList()
-		r := make([]v1.NodeSelectorRequirement, len(req))
-		copy(r, req)
-		for key, values := range supported {
-			r = append(r, v1.NodeSelectorRequirement{Key: key, Operator: v1.NodeSelectorOpIn, Values: values})
-		}
-		requirements := v1alpha5.NewRequirements(r...)
-		if err := requirements.Compatible(constraints.Requirements); err == nil {
-			return true
+		if !canScheduleOnExistingNode {
+			n := NewTheoreticalNode(constraints, instanceTypes)
+			n.Add(pod)
+			nodes = append(nodes, n)
```

Comment:
would be kind of cool if we made Add return the node object (builder pattern) to enable 
`nodes = append(nodes, NewTheoretricalNode(constraints, instanceTypes).Add(pod))`

Reply:
I like the second better, done.